### ISSUE TITLE
raiseProvisioningFailure requires undocumented argument with mandated wording 

### ISSUE THREAD

**Issue by nmalkin** [mozilla/persona#2339]
I think I just discovered an issue with either our implementation of provisioning for IdPs, or with our documentation, or with me.

For background, when we load the provisioning page from the IdP, we ask them to check the user's identity and, if in cases where authentication is needed, [call navigator.id.raiseProvisioningFailure()](https://developer.mozilla.org/en-US/docs/Persona/Implementing_a_Persona_IdP#The_Provisioning_Page).

In our example code, we call it like this: `navigator.id.raiseProvisioningFailure('user is not authenticated as target user')`, but our documentation doesn't say anything about the argument to the function.

Now take a look [at this line](https://github.com/mozilla/browserid/blob/2143dfa7ec540e2f3d39e3b7058753eea5467afd/resources/static/common/js/user.js#L508) in BrowserID code. It seems to be that we **require** the message to be _user is not authenticated as target user_.

This is supported by some experimental IdP code that I wrote (not on Github yet), where any provisioning-failure message not matching the one above results in the login process failing with the error, "_Action: Checking Address Info._"

Is there some deep reason for this? Is it somehow part of the protocol? If so, we should definitely update our documentation. Otherwise, I see no reason at all for the second part of that _if_-statement to be there.

(On a related note, why do we even have the `error.msg` field if we never use it?)

Note that this issue affects at least one of our IdPs: eyedee.me. In two places, it [calls raiseProvisioningFailure](https://github.com/lloyd/eyedee.me/blob/master/static/browserid/provision.html#L41) with a "forbidden" message.
But hostedpersona [seems to use](https://github.com/ozten/hostedpersona/blob/master/views/provisioning.ejs#L15) the (single) correct message everywhere, so maybe @ozten knows something I don't.

**Comment by ozten** [comment]
+1 this bite me too

**Comment by ozten** [comment]
https://github.com/mozilla/browserid/issues/1207

**Comment by nmalkin** [comment]
(Repeating for internal linking:)

This issue affected [someone on the mailing list](http://groups.google.com/group/mozilla.dev.identity/browse_thread/thread/ea00cf0bac16b1ca/3e0f360e37a9a24b?lnk=gst&q=raiseprovisioningfailure#3e0f360e37a9a24b), was filed as #1182, where @ozten [discovered](https://github.com/mozilla/browserid/issues/1182#issuecomment-4146610) what I just did, and filed #1207, which is open but has had no discussion.

**Comment by callahad** [comment]
> If so, we should definitely update our documentation.

Done.

> Otherwise, I see no reason at all for the second part of that if-statement to be there.

Not being too familiar with core, it looks like we're trying to distinguish between two cases:
- We can't continue provisioning because the user isn't authenticated.
- We can't continue provisioning because of an unrelated error.

In the first case, we recover by showing the user the `authentication` page. In the second case, we show a hard failure.

If this understanding is correct, I'd propose keeping the existing functionality (backwards compatibility), but adding a new callback for explicitly signaling that the user needs to authenticate. Maybe `navigator.id.raiseAuthenticationNeeded`?

Also: It might be worth checking for something similar in the implementation of `navigator.id.raiseAuthenticationFailure`.

**Comment by warner** [comment]
I wouldn't even describe the "authentication needed" situation as an exception at all: it happens at least once per user per browser, and requires special handling, so it's a normal occurrence, and "exception" isn't the right word for it. As such, I wouldn't use the word "raise" in the API for it. Maybe have the provisioning frame terminate by calling something named `navigator.id.authenticationNeeded()`?

**Comment by callahad** [comment]
Someone make a proposal and let's get this wart fixed.

Even if we further refine the API soon after Beta 2, I'd still be really happy if we fixed this one little nit right now.

**Comment by lloyd** [comment]
on it.

**Comment by lloyd** [comment]
read code, read proposals, I like what @warner suggests:

**Proposed Immediate fix:**
1. introduce `navigator.id.authenticationNeeded()`
2. for compat, make `.raiseProvisioningFailure()` with a string of `user is not authenticated as target user` equivalent to calling `.authenticationNeeded()`

**Proposed future UX improvement:**
3. If any other failure is raised during initial provisioning attempt, let's display a user facing error message "we're sorry, <domain> cannot log you in right now <actual error hidden behind click> <button>Try a different email address</button>"

NOTE: I don't know what the failure mode actually is today for first-time prov failure, but I'll set that up.

**Comment by shane-tomlinson** [comment]
Can the API be simplified using node-esque syntax:

```
function checkActiveSession(email, done) {   
   // do some checking. True if the user has a session, false if not    
   done(err, true || false);
}

function generateCertificate(email, publicKey, certDuration, done) {   
    // generate and sign key server side   
   done(err, certificate);
}

navigator.id.provision(checkActiveSession, generateCertificate);
```

**Comment by callahad** [comment]
@shane-tomlinson That looks like an awesome start on a post-beta-2 refinement :)

**Comment by lloyd** [comment]
**Refined proposed immediate fix:**
1. any invocation of `.raiseProvisioningFailure()` during the first attempt (pre-auth) causes authentication url to be loaded
2. any invocation of `.raiseProvisioningFailure()` during second attempt (post-auth) causes us to redirect back to email entry

theory is the IdP tells the user what the authentication problem is, and us displaying an error screen would only be confusing.  This addresses the main API pain point in a backcompat way and should only mean we remove a couple lines of code.

Later there are tons of opportunities to improve the idp api.

**Comment by lloyd** [comment]
The above PR is minimalistic, makes the behavioral change mentioned above, and awaits the thoughts of a front-end hero.

**Comment by jaredhirsch** [comment]
fixed by #3081. closing
